src/ott/neural/solvers/expectile_neural_dual.py

```diff
@@ -0,0 +1,655 @@
+
+  @classmethod
+  def _padder(cls, dim: int) -> jnp.ndarray:
+    return jnp.ones((1, dim))
+
+
+class PotentialTrainState(train_state.TrainState):
+  r"""Adds information about the model's value and gradient to the state.
+
+  The gradient may differ from composition of jax.grad and value functions.
+
+  Args:
+  potential_value_fn: the potential's value function.
+  potential_gradient_fn: the potential's gradient function.
+  """
+
+  potential_value_fn: Callable[[frozen_dict.FrozenDict[str, jnp.ndarray]],
+                               PotentialValueFn_t] = struct.field(
+                                   pytree_node=False
+                               )
+
+  potential_gradient_fn: Callable[[frozen_dict.FrozenDict[str, jnp.ndarray]],
+                                  PotentialGradientFn_t] = struct.field(
+                                      pytree_node=False
+                                  )
+
+
+class PotentialModelWrapper(nn.Module):
+  r"""Wrapper class for the neural models.
+
+  Implements a potential value or a vector field.
+  """
+
+  is_potential: bool  # Implements a potential value or a vector field.
+  add_l2_norm: bool  # If true, l2 norm is added to the potential.
+  model: nn.Module  # The potential model.
+
+  def apply(
+      self, params: frozen_dict.FrozenDict[str, jnp.ndarray], x: jnp.ndarray
+  ) -> jnp.ndarray:
+    """Apply the potential model, optionally add squared Euclidean norm."""
+    z: jnp.ndarray = self.model.apply({"params": params}, x)
+
+    if self.is_potential:
+      z = z.squeeze()
+
+    if self.is_potential and self.add_l2_norm:
+      z = z + 0.5 * jnp.dot(x, x)
+    if not self.is_potential and self.add_l2_norm:
+      z = z + x
+
+    return z
+
+  def potential_value_fn(
+      self, params: frozen_dict.FrozenDict[str, jnp.ndarray]
+  ) -> PotentialValueFn_t:
+    """Return a function giving the value of the potential."""
+    assert (
+        self.is_potential is True
+    ), "A model should be potential to get potential_value_fn."
+    return lambda x: self.apply(params, x)
+
+  def potential_gradient_fn(
+      self, params: frozen_dict.FrozenDict[str, jnp.ndarray]
+  ) -> PotentialGradientFn_t:
+    """A vector function or gradient of the potential."""
+    if self.is_potential:
+      return jax.grad(self.potential_value_fn(params))
+    return lambda x: self.apply(params, x)
+
+  def create_train_state(
+      self,
+      rng: jax.Array,
+      optimizer: optax.GradientTransformation,
+      input: Union[int, Tuple[int, ...]],
+      **kwargs: Any,
+  ) -> PotentialTrainState:
+    """Create initial training state."""
+    params = self.model.init(rng, jnp.ones(input))["params"]
+
+    return PotentialTrainState.create(
+        apply_fn=self.apply,
+        params=params,
+        tx=optimizer,
+        potential_value_fn=self.potential_value_fn,
+        potential_gradient_fn=self.potential_gradient_fn,
+        **kwargs
+    )
+
+
+class MLP(nn.Module):
```

Comment:
Am thinking of the following:
with the wrapper in mind, we could move this to `ott.neural.networks.potentials` (including the wrapper) and remove the `PotentialMLP` (both of the 2 classes you define are more general) it seems.
Pinging @marcocuturi @Algue-Rythme for opinion.

Reply:
I have left this class in the same place for compatibility, but support the idea to replace `PotentialMLP` in the future.